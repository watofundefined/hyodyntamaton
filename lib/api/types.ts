import { ApiError } from 'lib/http'
import { Venue } from 'lib/types'

export interface VenuesSearchRequest {
  lat: number
  lng: number
  radius: number
}

export interface VenuesSearchResponse {
  venues?: Venue[]
  error?: ApiError
}
