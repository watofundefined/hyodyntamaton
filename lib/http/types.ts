// eslint-disable-next-line no-restricted-syntax
import { AxiosRequestConfig, AxiosResponse, AxiosInterceptorManager } from 'axios'

export interface ApiReqConfig<TParams> extends AxiosRequestConfig {
  params?: TParams
}

export type Result<TData, TError = null> =
  | { status: number; error: null; data: TData }
  | { status: number; error: TError; data: null }

export type ApiError = { code: string; message: string }

export type ApiResult<TData> = Promise<Result<TData, ApiError>>

// Subset of AxiosInstance - expose only what's needed
export interface ApiClient {
  get<TReq, TData>(url: string, config?: ApiReqConfig<TReq>): ApiResult<TData>
  interceptors: {
    request: AxiosInterceptorManager<AxiosRequestConfig>
    response: AxiosInterceptorManager<AxiosResponse>
  }
}
