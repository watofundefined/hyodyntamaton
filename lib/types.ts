export type QueryDict<T> = { [key in keyof T]: string | string[] }

export interface GeoLocation {
  lat: number
  lng: number
}

export interface Venue {
  ids: VenueIds
  name: string
  url: string
  location: GeoLocation
  categories: VenueCategory[]
}

export interface VenueIds {
  foursquareId: string
  untappedId?: number
}

export type VenueCategory =
  | 'beer-bar'
  | 'beer-garden'
  | 'beer-shop'
  | 'brewery'
  | 'cidery'
  | 'other'
