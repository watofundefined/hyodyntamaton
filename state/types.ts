import { UserState } from './user/reducer'
import { VenuesState } from './venues/reducer'
import { BeersState } from './beers/reducer'

export type Action<TType extends string = string, TPayload = undefined> = {
  type: TType
  payload?: TPayload
}

export interface AppState {
  user: UserState
  venues: VenuesState
  beers: BeersState
}

export type { UserState } from './user/reducer'
export type { VenuesState } from './venues/reducer'
export type { BeersState } from './beers/reducer'
