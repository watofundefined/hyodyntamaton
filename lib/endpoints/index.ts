export * from './foursquare'
export * from './untappd'
