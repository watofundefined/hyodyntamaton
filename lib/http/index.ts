import axios from 'axios'
import configureInterceptors from './configure-interceptors'
import { ApiClient } from './types'

const client: ApiClient = axios.create()
configureInterceptors(client)

export * from './types'
export { client }
