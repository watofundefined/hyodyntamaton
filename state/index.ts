import { combineReducers } from 'redux'
import user from './user/reducer'
import venues from './venues/reducer'
import beers from './beers/reducer'
import { AppState } from './types'

function getInitialState(state: Partial<AppState> = {}): AppState {
  return {
    user: user.defaultState(),
    venues: venues.defaultState(),
    beers: beers.defaultState(),
    ...state,
  }
}

const reducer = combineReducers<AppState>({
  user: user.reducer,
  venues: venues.reducer,
  beers: beers.reducer,
})

export { reducer, getInitialState }
export * from './types'
export { default as UserActions } from './user/actions'
export { default as VenuesActions } from './venues/actions'
export { default as BeersActions } from './beers/actions'
