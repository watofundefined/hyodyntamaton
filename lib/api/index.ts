import { foursquareLookup, venueInfo, venueCheckins } from 'lib/endpoints'
import { ApiResult, client } from 'lib/http'
import { VenueIds } from 'lib/types'
import { VenuesSearchRequest, VenuesSearchResponse } from './types'
import { UtVenueInfoResponse } from 'lib/endpoints/untappd/venue-info.types'
import {
  UtVenueCheckinsRequest,
  UtVenueCheckinsResponse,
} from 'lib/endpoints/untappd/venue-checkins.types'

export default {
  venues: {
    search: (req: VenuesSearchRequest): ApiResult<VenuesSearchResponse> => {
      return client.get('/api/venues-search', {
        params: req,
      })
    },
    foursquareIdToUntappdId: (id: string, untappedToken: string): ApiResult<VenueIds> => {
      return foursquareLookup(id, { access_token: untappedToken })
    },
    info: (untappdId: number, untappedToken: string): ApiResult<UtVenueInfoResponse> => {
      return venueInfo(untappdId, { access_token: untappedToken })
    },
    checkins: (
      id: number,
      req: UtVenueCheckinsRequest
    ): ApiResult<UtVenueCheckinsResponse> => {
      return venueCheckins(id, req)
    },
  },
}

export * from './types'
