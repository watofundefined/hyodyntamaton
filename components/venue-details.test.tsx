import api from 'lib/api'
import { Result } from 'lib/http'
import { VenueIds } from 'lib/types'
import { AppState } from 'state'
import { render } from 'test/test-utils'
import VenueDetails from './venue-details'

jest.mock('lib/api')

beforeAll(() => {
  ;(api.venues.foursquareIdToUntappdId as jest.Mock).mockImplementation((fsId) => {
    return new Promise<Result<VenueIds>>((resolve) =>
      resolve({
        data: {
          foursquareId: fsId,
          untappedId: 143241234,
        },
        status: 200,
        error: null,
      })
    )
  })
})

beforeEach(() => jest.clearAllMocks())

describe('VenueDetails component', () => {
  const fsId = 'abcd1234'
  const userToken = 'token'

  it('matches snapshot', async () => {
    const { asFragment } = render(<VenueDetails venueFsId={fsId} />, {
      state: makeState(userToken, fsId),
    })

    expect(asFragment()).toMatchSnapshot()
  })

  it("fetches the Untappd Id if it's missing", async () => {
    render(<VenueDetails venueFsId={fsId} />, { state: makeState(userToken, fsId) })

    expect(api.venues.foursquareIdToUntappdId).toHaveBeenCalledTimes(1)
    expect(api.venues.foursquareIdToUntappdId).toHaveBeenCalledWith(fsId, userToken)
  })

  it("doesn't fetch the Untappd Id if it's already populated", async () => {
    render(<VenueDetails venueFsId={fsId} />, { state: makeState(userToken, fsId, 1234) })

    expect(api.venues.foursquareIdToUntappdId).not.toHaveBeenCalled()
  })
})

function makeState(token: string, fsId: string, utId?: number): Partial<AppState> {
  return {
    user: { loggedIn: true, token },
    venues: {
      items: [
        {
          ids: {
            foursquareId: fsId,
            untappedId: utId,
          },
          name: 'Dummy Name',
          categories: ['beer-bar'],
          location: { lat: 123, lng: 123 },
          url: 'http://example.com',
        },
      ],
    },
  }
}
