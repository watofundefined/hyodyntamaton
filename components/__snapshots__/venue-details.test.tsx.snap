// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VenueDetails component matches snapshot 1`] = `
<DocumentFragment>
  <div
    __self="[object Object]"
    __source="[object Object]"
    id="__next"
  >
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="gmap-container"
    />
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="venue-details"
    >
      <h2
        __self="[object Object]"
        __source="[object Object]"
        class="venue-name"
      >
        Dummy Name
        <span
          __self="[object Object]"
          __source="[object Object]"
          class="venue-category"
        >
          Beer Bar
        </span>
      </h2>
      <section
        __self="[object Object]"
        __source="[object Object]"
        class="venue-details-checkins"
        style="max-height: -115px;"
      >
        <header
          __self="[object Object]"
          __source="[object Object]"
        >
          Latest feed:
        </header>
        <ol
          __self="[object Object]"
          __source="[object Object]"
        />
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="checkins-bottom-drawer"
        >
          0
        </div>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
